bundles/org.eclipse.swt/Eclipse SWT Program/win32/org/eclipse/swt/program/Program.java

```diff
@@ -418,15 +396,42 @@ public ImageData getImageData (int zoom) {
 		}
 	}
 	TCHAR lpszFile = new TCHAR (0, fileName, true);
-	long [] phiconSmall = new long[1], phiconLarge = null;
-	OS.ExtractIconEx (lpszFile, nIconIndex, phiconLarge, phiconSmall, 1);
-	if (phiconSmall [0] == 0) return null;
-	Image image = Image.win32_new (null, SWT.ICON, phiconSmall [0], initialNativeZoom);
+	long [] hIcon = new long[1];
+	int size = OS.GetSystemMetricsForDpi(OS.SM_CXSMICON, DPIUtil.mapZoomToDPI(zoom));
+	OS.SHDefExtractIcon (lpszFile, nIconIndex, 0, hIcon, null, size);
+	if (hIcon [0] == 0) {
+		return getImageDataUsingExtension(zoom);
+	}
+	Image image = Image.win32_new (null, SWT.ICON, hIcon [0], zoom);
 	ImageData imageData = image.getImageData (zoom);
 	image.dispose ();
 	return imageData;
 }
 
+private ImageData getImageDataUsingExtension(int zoom) {
+	if (extension != null) {
+		int initialNativeZoom = getPrimaryMonitorZoomAtStartup();
```

Comment:
Please keep the existing comment for why we use primary monitor zoom at startup. Otherwise it will be even harder to comprehend for other / in the future.